File: nn/src/test/scala/com/intel/webscaleml/nn/nn/performTest/SpatialMaxPoolingPerform.lua
--
-- Created by: yao on 6/6/16
--

require 'nn'

forwardIterations = 10
backwardIterations = 10
batchSize = 100
nInputPlane = 3
height = 100
weight = 512
seed = 100
allTestCases = {
    {kW = 3, kH = 3, dW = 2, dH = 2}, -- AlexNet, GoogleNet
    {kW = 2, kH = 2, dW = 2, dH = 2}, -- Cifar, CifarLocal
    {kW = 3, kH = 3, dW = 1, dH = 1}, -- GoogleNet
    {kW = 3, kH = 3, dW = 3, dH = 3}, -- MNIST
}
torch.manualSeed(seed)
file = io.open("TorchPerform.csv", "a")
io.output(file)

input = torch.Tensor(batchSize, nInputPlane, height, weight)
input:apply(function()
    i = torch.uniform(0, 1)
    return i
end)

for k, testCase in pairs(allTestCases) do
    local model = nn.SpatialMaxPooling(testCase.kW, testCase.kH, testCase.dW, testCase.dH)
    local timer = torch.Timer()
    for j = 1, forwardIterations do
        model:forward(input)
    end
    local timeSpend = timer:time().real / forwardIterations
    file:write("SpatialMaxPooling(" .. testCase.kW .."," ..
        testCase.kH .. "," .. testCase.dW .. "," .. testCase.dH .. ")forward:" .. string.format("%.6f", timeSpend) * 1000 .. "\n")
end

for k, testCase in pairs(allTestCases) do
    local model = nn.SpatialMaxPooling(testCase.kW, testCase.kH, testCase.dW, testCase.dH)
    local output = model:forward(input)
    local grad = output:clone():apply(function()
        i = torch.uniform(-2, 2)
        return i
    end)
    local timer = torch.Timer()
    for j = 1, backwardIterations do
        model:backward(input, grad)
    end
    local timeSpend = timer:time().real / backwardIterations
    file:write("SpatialMaxPooling(" .. testCase.kW .."," ..
            testCase.kH .. "," .. testCase.dW .. "," .. testCase.dH .. ")backward:" .. string.format("%.6f", timeSpend) * 1000 .. "\n")
end

file:close()

File: dl/src/test/scala/com/intel/analytics/dllib/lib/nn/performTest/SpatialMaxPoolingPerform.scala
package com.intel.webscaleml.nn.nn.performTest

import java.io.{BufferedWriter, File, FileWriter}

import com.intel.webscaleml.nn.nn.SpatialMaxPooling
import com.intel.webscaleml.nn.tensor.{RandomGenerator, torch}
import org.scalatest.{BeforeAndAfter, FlatSpec}

/**
  * Created by yao on 5/19/16.
  */
class SpatialMaxPoolingPerform extends FlatSpec with BeforeAndAfter{
  val ForwardIterations = PerformTestData.forwardIterations
  val BackwardIterations = PerformTestData.backwardIterations
  val BatchSize = 100
  val NInputPlane = 3
  val Height = 100
  val Width = 512
  val Seed = 100
  val allTestCases = List(
    TestCase(3, 3, 2, 2, 40, 20), // AlexNet, GoogleNet
    TestCase(2, 2, 2, 2, 30, 25), // Cifar, CifarLocal
    TestCase(3, 3, 1, 1, 130, 25), //GoogleNet
    TestCase(3, 3, 3, 3, 22, 20) // MNIST
  )

  val input = torch.Tensor[Float](BatchSize, NInputPlane, Height, Width).rand()

  for (testCase <- allTestCases) {
    val model = new SpatialMaxPooling[Float](testCase.kW, testCase.kH, testCase.dW, testCase.dH)

    "A SpatialMaxPoolingPerform" should s" forward with a good performance with parameters" +
      s": ${testCase.kW}, ${testCase.kH}, ${testCase.dW}, ${testCase.dH}," in {
      if (!TestUtils.isRun())
        cancel("Performance regression tests are forbidden")
      val timeMillis = TestUtils.testModuleForwardPerform(input, model, ForwardIterations)
      assert(timeMillis < testCase.ForwardTimeoutMillis)
    }
  }

  for (testCase <- allTestCases) {
    val model = new SpatialMaxPooling[Float](testCase.kW, testCase.kH, testCase.dW, testCase.dH)
    val output = model.forward(input)
    val grad = output.clone().rand()

    "A SpatialMaxPoolingPerform" should s" backward with a good performance with parameters" +
      s": ${testCase.kW}, ${testCase.kH}, ${testCase.dW}, ${testCase.dH}," in {
      if (!TestUtils.isRun())
        cancel("Performance regression tests are forbidden")
      val timeMillis = TestUtils.testModuleBackwardPerform(input, grad, model, BackwardIterations)
      assert(timeMillis < testCase.BackwardTimeoutMillis)
    }
  }

  case class TestCase(kW: Int, kH: Int, dW: Int, dH: Int, ForwardTimeoutMillis: Int, BackwardTimeoutMillis: Int)
}


